8348663: [AIX] clang pollutes the burned-in library search paths of the generated executables

JoKern65 opened the pull request:
The clang compiler which we use since jdk21 pollutes the burned-in library search paths of the generated executables.
e.g. a dump -X64 -H ... /jdk-23.0.1+11-jre/lib/libnio.so

amongst others produces the following lines

                        ***Import File Strings***
INDEX PATH BASE MEMBER
0 /opt/IBM/openxlC/17.1.1/../../../../usr/lpp/xlC/lib:/opt/IBM/openxlC/17.1.1/lib:/opt/IBM/openxlC/17.1.1/bin/../../../../../usr/lpp/xlC/lib:/usr/opt/zlibNX/lib:/opt/freeware/lib/pthread:/opt/freeware/lib:/usr/lib

The long string is the burned in library search path which will be used after searching with LIBPATH envvar.
You see there are some compiler paths which customers will not have; also /opt/freeware is unwanted.
A proper executable should only have the burned in default search path
/usr/lib:lib

To ensure this we want to call the linker with the flag
-fuse-ld=our-wrapper-script-cleaning-LIBPATH- and-calling-original-ld-afterwards
This script we want to deliver in make/scripts/aix/ld.sh
just containing
#!/bin/bash
unset LIBPATH
exec /usr/bin/ld "$@"
return $?

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8348663](https://bugs.openjdk.org/browse/JDK-8348663): [AIX] clang pollutes the burned-in library search paths of the generated executables (**Bug** - P4)


### Reviewers
 * [Magnus Ihse Bursie](https://openjdk.org/census#ihse) (@magicus - **Reviewer**)
 * [Matthias Baesken](https://openjdk.org/census#mbaesken) (@MBaesken - **Reviewer**) Review applies to [46a2e67e](https://git.openjdk.org/jdk/pull/23313/files/46a2e67ead54ed98da56e2e8ecb1e39af8f98951)
 * [Christoph Langer](https://openjdk.org/census#clanger) (@RealCLanger - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/23313/head:pull/23313` \
`$ git checkout pull/23313`

Update a local copy of the PR: \
`$ git checkout pull/23313` \
`$ git pull https://git.openjdk.org/jdk.git pull/23313/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 23313`

View PR using the GUI difftool: \
`$ git pr show -t 23313`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/23313.diff">https://git.openjdk.org/jdk/pull/23313.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/23313#issuecomment-2621158780)
</details>

JoKern65 commented:
@Magicus Hi Magnus, I'm faced with the problem that I cannot check in an executable script. On the other side the compilers -fuse_ld parameter needs an executable script. Is there a way to circumvent this.
The only fallback I'm aware of would be to keep the script local and just commit a configure parameter which provides the path to the local script setting then -fuse-ld accordingly.
Are there other solutions possible?

magicus commented:
This look like a very hacky workaround overall. Can't you just unset LIBPATH in spec.gmk? Or just before linking? Is this a documented behavior of the linker on AIX, or a bug?

JoKern65 commented:
Unfortunately, there is no other way. The LIBPATH envvar during our build is already unset. The compiler itself, when called as the Linker is setting some paths into LIBPATH before he calls the real Linker by himself. This is the only way to get rid of the unwanted burned in paths. This was an advice from IBM to do so as a standard trick.
I now checked in the script without x-Bit. Then during configure call it is copied to the build directory, where the x-Bit is set. -fuse-ld= Parameter points now to this copied version.
If you are reluctant with this solution, we additionally can make it dependent from a new configure flag which we would set and nobody else.
By the way, if we try to address this as a possible bug and this is very unlikely accepted by IBM, it will take at least half a year to get a change. And because this would change the behavior of the compiler I'm pretty sure this will only be included in a new Compiler not yet shipped.

magicus commented:
I still think you should report it as a bug, but I accept that we need to have a workaround in the meantime (or forever, if it doesn't get fixed).

However, you need to shape up the patch a bit first.

* `$OUTPUTDIR` points to the current configure dir, use that instead.
* You should use the detected versions of cp and chmod as $CP and $CMHOD.
* You should not copy files in flags-ldflags.m4, that parts needs to find a better home.
* The `return $?` part in the script is useless since `exec` will make the ld process replace the shell script.

JoKern65 commented:
I followed all of Magnus proposals:
I used `$OUTPUTDIR` instead of `${TOPDIR}/build/${CONF_NAME}`.
I replaced `cp` and `chmod` by `$CP` and `$CMHOD`.
I moved the copy and mode bit change from `flags-ldflags.m4` to `BASIC_POST_CONFIG_OUTPUT` in `basic.m4` .
I removed the useless `return $?` from my script.

magicus commented on make/autoconf/basic.m4:
```diff
@@ -624,4 +624,11 @@ AC_DEFUN_ONCE([BASIC_POST_CONFIG_OUTPUT],
+
+  # Copy the linker wrapper script for AIX' clang and make it executable
+  if test "x$TOOLCHAIN_TYPE" = xclang && test "x$OPENJDK_TARGET_OS" = xaix &&
+     test -e "${TOPDIR}/make/scripts/aix/ld.sh"; then
```
You can assume that checked-in files exist and do not need to test for them.

magicus commented on make/autoconf/basic.m4:
```diff
@@ -624,4 +624,11 @@ AC_DEFUN_ONCE([BASIC_POST_CONFIG_OUTPUT],
+  # Copy the linker wrapper script for AIX' clang and make it executable
+  if test "x$TOOLCHAIN_TYPE" = xclang && test "x$OPENJDK_TARGET_OS" = xaix &&
+     test -e "${TOPDIR}/make/scripts/aix/ld.sh"; then
+    $CP -f "${TOPDIR}/make/scripts/aix/ld.sh" "$OUTPUTDIR/ld.sh"
```
```suggestion
    $CP -f "$TOPDIR/make/scripts/aix/ld.sh" "$OUTPUTDIR/ld.sh"
```

We normally prefer to use $ without {} unless they are necessary, but in some parts of the code the ${} style is prevalent for historical reasons, and in those places it is better to keep with that style. But for code like this, you should stick to the {}-less style.

JoKern65 replied:
Done

magicus commented on make/autoconf/basic.m4:
```diff
@@ -624,4 +624,10 @@ AC_DEFUN_ONCE([BASIC_POST_CONFIG_OUTPUT],
   # Make the compare script executable
   $CHMOD +x $OUTPUTDIR/compare.sh
+
+  # Copy the linker wrapper script for AIX' clang and make it executable
```
```suggestion
  # Copy the linker wrapper script for clang on AIX and make it executable
```

magicus commented:
Just one comment fix and then this is okay. Also please verify that the latest version still works on AIX.

JoKern65 commented:
All done and also latest version still works on AIX.

magicus reviewed (approved):
Looks good now. Thanks!

RealCLanger commented:
Looks good to me, too. However, you should update the copyright years in the modified files and also add a license header in the newly added script (like in other .sh files in that directory but with SAP copyright).

JoKern65 commented:
Copyrights modified/added.

RealCLanger commented on make/scripts/aix/ld.sh:
```diff
@@ -0,0 +1,28 @@
+#!/bin/bash
+#
+# Copyright (c) 2025, Oracle and/or its affiliates. All rights reserved.
```
You can leave out Oracle here.

JoKern65 replied:
done

RealCLanger reviewed (approved):
Cloes to perfect. 😄

JoKern65 commented:
/integrate
